src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -0,0 +1,434 @@
+            ?: return fallback
+
+        val fnTrait = typeParameter.getTraitBoundsTransitively()
+            .find { it.element.isAnyFnTrait }
+            ?: return fallback
+
+        return fnTrait.asFunctionType?.substitute(containingFunctionType.typeParameterValues)
+            ?: fallback
+    }
+
+    private fun inferArrayType(expr: RsArrayExpr): Ty {
+        val (elementType, size) = if (expr.semicolon != null) {
+            val elementType = expr.initializer?.ty ?: return TySlice(TyUnknown)
+            val size = calculateArraySize(expr.sizeExpr) ?: return TySlice(elementType)
+            elementType to size
+        } else {
+            val elements = expr.arrayElements
+            if (elements.isNullOrEmpty()) return TySlice(TyUnknown)
+
+            var elementType: Ty = TyUnknown
+            // '!!' is safe here because we've just checked that elements isn't null
+            for (e in elements!!) {
+                elementType = getMoreCompleteType(e.ty, elementType)
+            }
+            elementType to elements.size
+        }
+        return TyArray(elementType, size)
+    }
+
+    private fun getMoreCompleteType(t1: Ty, t2: Ty): Ty {
+        return when {
+            t1 is TyUnknown -> t2
+            t1 is TyInteger && t2 is TyInteger && t1.isKindWeak -> t2
+            t1 is TyFloat && t2 is TyFloat && t1.isKindWeak -> t2
+            else -> t1
+        }
+    }
+
+    private fun inferStructTypeArguments(literal: RsStructLiteral): Substitution =
+        inferFieldTypeArguments(literal.structLiteralBody.structLiteralFieldList)
+
+    private fun inferFieldTypeArguments(fieldExprs: List<RsStructLiteralField>): Substitution {
+        val argsMapping = mutableMapOf<TyTypeParameter, Ty>()
+        fieldExprs.forEach { field -> field.expr?.let { expr -> addTypeMapping(argsMapping, field.type, expr) } }
+        return argsMapping
+    }
+
+    private fun mapTypeParameters(argDefs: Iterable<Ty>, argExprs: Iterable<RsExpr>): Substitution {
+        val subst = mutableMapOf<TyTypeParameter, Ty>()
+        argExprs.zip(argDefs).forEach { (expr, type) -> addTypeMapping(subst, type, expr) }
+        return subst
+    }
+
+    private fun addTypeMapping(argsMapping: TypeMapping, fieldType: Ty?, expr: RsExpr) =
+        fieldType?.canUnifyWith(expr.ty, expr.project, argsMapping)
+
+    private fun inferArithmeticBinaryExprType(expr: RsBinaryExpr, op: ArithmeticOp): Ty {
+        val lhsType = expr.left.ty
+        val rhsType = expr.right?.ty ?: TyUnknown
+        return findArithmeticBinaryExprOutputType(expr.project, lhsType, rhsType, op)
+    }
+}
+
+fun collectBindingsNullable(pattern: RsPat?, baseType: Ty): Map<RsPatBinding, Ty> {
+    return if (pattern != null) collectBindings(pattern, baseType) else emptyMap()
+}
+
+private object TypeInferenceRecursionGuard {
+    private val local: ThreadLocal<Boolean> = ThreadLocal.withInitial { false }
+
+    inline fun guarded(action: () -> Unit) {
```

Comment:
Let's write a short docstring here explaining that this is for assertions only. Also, looks like we don't need `TypeInferenceRecursionGuard` wrapper object?